### ISSUE TITLE
Compatibility issue with faiss-cpu wheels on arm64 vs x86

### ISSUE THREAD

**Issue by NikhilSehgal123** [kyamagu/faiss-wheels#74]
**Describe the bug**
FAISS has different wheels for arm64 chips & Intel/AMD chips. If you vectorise a document using arm64, you should use FAISS to interact with this document on a arm64 machine.

So right now, my mac is arm64 and I vectorised docs on my mac using the arm64 wheel. However in the Azure functions, it attempts to use the Intel wheel which is slightly different. Intel/AMD chips use something called AVX2 support in FAISS (no idea what it actually is but it's different to the way arm64 machines use FAISS).

**How can this be solved?**
Normalise the wheels across all architectures to solve this compatibility issue problem - happy to help?

**Comment by kyamagu** [comment]
[avx2](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions) is a CPU instruction set. I have no idea what is the binary compatibility to the saved indices across different architectures. Ask the question in [the official repository](https://github.com/facebookresearch/faiss).

**Comment by max-addison** [comment]
Yep, I have this exact issue. I've deployed to AWS EKS but built the vectorstore on my mac locally and am getting errors. Would be great to get this fixed! @NikhilSehgal123 if you post this issue in the official repository, please post the link here, and i'll +1 and comment! 🙏

**Comment by jpzhangvincent** [comment]
Ya it would be great to fix this issue soon if possible!

**Comment by NikhilSehgal123** [comment]
Hey guys @max-addison @jpzhangvincent, I know this issue is causing some pain for many people, so another simple workaround is to just upload your vectorstores to Pinecone or another cloud hosted vector DB to avoid this type of issue.

**Comment by Alex-Wenner-FHR** [comment]
I experienced this same issue when using LangChain.

[here](https://github.com/hwchase17/langchain/blob/6cdd4b5edca511b0015f1b39102225fe638d8359/langchain/vectorstores/faiss.py#L23) there is the following code:
```python
Import faiss if available, otherwise raise error.
    If FAISS_NO_AVX2 environment variable is set, it will be considered
    to load FAISS with no AVX2 optimization.
```

When I explicitly did not set `FAISS_NO_AVX2` this error subsides.  Previously I was experiencing the same thing mentioned in this thread. The faiss index that I am working with today, was created on a windows machine and running successfully on a linux machine. So it seems like loading FAISS with AVX2 optimization resolved the issue for me. Not sure this is helpful to everybody, but hopefully it will help somebody 😄

**Comment by aiakubovich** [comment]
@Alex-Wenner-FHR 

I was trying `FAISS_NO_AVX2` but it does not work inside of docker container for some reason (getting `TypeError: IndexFlatCodes.add() missing 1 required positional argument: 'x'` error). I opened issue: https://github.com/langchain-ai/langchain/issues/8857  . Please let me know if you know why I am getting that error.

**Comment by Alex-Wenner-FHR** [comment]
@aiakubovich I actually ran into this as well. I am trying to recall what I had done... I know I had troubles with pathing and being able to find the `index.faiss` in the `faiss_index` folder. I do believe I had to recreate the Faiss_Index on the same type of machine I was trying to inference... for example, originally it was on Windows and trying to be ran on a linux machine, I had to recreate it on a linux machine to run it on the linux machine.

**Comment by aiakubovich** [comment]
@Alex-Wenner-FHR Thanks for your response. I thought that `FAISS_NO_AVX2` is option that allows to use same embeddings for both Windows and Linux machines. Without this option I indeed have to rerun embeddings on different machines. With this option I unable to run embeddings in Linux container at all. No idea why.

**Comment by aboutmydreams** [comment]
Same problem to run on linux but not on m2/m1 mac

https://github.com/langchain-ai/chat-langchain/issues/61

**Comment by kaixxx** [comment]
I am also struggling with this problem for hours...
Now I've discovered something very strange: If I go directly into the installation folder of `faiss-cpu` on my system and delete `_swigfaiss_avx2.cp312-win_amd64.pyd`, my app will use the non-AVX2 version and finally create an index which can also be read on macOS. 

Like many others, I am working with Langchain, which uses the following code to select between AVX2 and non-AVX2 import:
```
if no_avx2: 
    from faiss import swigfaiss as faiss
else:
    import faiss
```
[Source](https://python.langchain.com/v0.2/api_reference/_modules/langchain_community/vectorstores/faiss.html#dependable_faiss_import)

However, even if `no_avx2` is `True`, AVX2 is used, resulting in an index which is not cross-compatible - unless I delete the AVX2-lib from my system as described above. Very strange. 
**Is this a problem with langchain or with faiss-cpu?** Any idea how to solve this?

[People over at langchain now reccommend going back to faiss-cpu 1.7.2 to solve this issue.](https://github.com/langchain-ai/chat-langchain/issues/61#issuecomment-1914426775) But this version is already 5 years old and only runs on Python 3.10...

**Comment by kaixxx** [comment]
Ok, it seems that I have finally found a solution!

There is a very poorly documented feature in faiss introduced with this commit: https://github.com/facebookresearch/faiss/commit/eefa39105eda498ab5fcd82ea0e11b18fd3fc0d7
It now uses the `FAISS_OPT_LEVEL` environment variable not only during build, but also during run time. The variable can be set to either `AVX2`, `AVX512` or an empty string to pick whatever x86_64 instruction set you want. Setting it to an empty string will produce a "generic" index that is also compatible with Mac or Linux machines:
```
import os
os.environ['FAISS_OPT_LEVEL'] = ''  # do this BEFORE importing any faiss library
```

Setting `os.environ['FAISS_NO_AVX2'] = '1'` in Langchain *will not work*.

**Comment by kyamagu** [comment]
Sure, I will include a note on `FAISS_OPT_LEVEL` in the README.md.

In any case, AVX2-built indices are not compatible with the generic extension. This is an issue that can only be fixed in the upstream.
